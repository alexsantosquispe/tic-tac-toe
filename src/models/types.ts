import type { ReactNode } from 'react';

export interface SvgIconProps {
  className?: string;
}
export const LANGUAGES = [
  { id: 'en', label: 'EN' },
  { id: 'es', label: 'ES' }
] as const;

export type LanguageTypes = (typeof LANGUAGES)[number]['id'];

export type SquareValueTypes = 'X' | 'O' | '';

export type CurrentPlayerType = Exclude<SquareValueTypes, ''>;

export const PLAYER_MODE = {
  SINGLE_PLAYER: 'singlePlayer',
  TWO_PLAYERS: 'twoPlayers'
} as const;

export type PlayerModeTypes = (typeof PLAYER_MODE)[keyof typeof PLAYER_MODE];

export const LEVELS = {
  EASY: 'easy',
  HARD: 'hard'
} as const;

export type LevelTypes = (typeof LEVELS)[keyof typeof LEVELS];

export type OptionType<T> = {
  value: T;
  component: ReactNode;
};
