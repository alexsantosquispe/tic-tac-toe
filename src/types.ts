export interface SvgIconProps {
  className?: string;
}
export const LANGUAGES = [
  { id: 'en', label: 'EN' },
  { id: 'es', label: 'ES' }
] as const;

export type LanguageTypes = (typeof LANGUAGES)[number]['id'];

export type SquareValueTypes = 'X' | 'O' | '';
