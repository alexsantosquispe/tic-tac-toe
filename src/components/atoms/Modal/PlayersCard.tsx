import type { ReactNode } from 'react';
import { BotIcon } from '../../../icons/BotIcon';

interface PlayersCardProps {
  isSinglePlayer?: boolean;
  playerIcon: ReactNode;
}

export const PlayersCard = ({
  isSinglePlayer,
  playerIcon
}: PlayersCardProps) => {
  return (
    <div className="flex items-center justify-between gap-8 px-8 py-4">
      {playerIcon}
      <span className="text-xl font-medium">VS</span>
      {isSinglePlayer ? <BotIcon className="size-12" /> : playerIcon}
    </div>
  );
};
