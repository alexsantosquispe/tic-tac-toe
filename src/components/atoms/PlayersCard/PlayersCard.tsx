import { memo } from 'react';
import { BotIcon } from '../../../icons/BotIcon';
import { PersonIcon } from '../../../icons/PersonIcon';

interface PlayersCardProps {
  isSinglePlayer?: boolean;
}

const PlayersCard = ({ isSinglePlayer = true }: PlayersCardProps) => {
  return (
    <div className="flex w-full items-center justify-between gap-8 px-8 py-6">
      <PersonIcon className="size-12" />

      <span className="text-xl font-medium">VS</span>

      {isSinglePlayer ? (
        <BotIcon className="size-12" />
      ) : (
        <PersonIcon className="size-12" />
      )}
    </div>
  );
};

export default memo(PlayersCard);
