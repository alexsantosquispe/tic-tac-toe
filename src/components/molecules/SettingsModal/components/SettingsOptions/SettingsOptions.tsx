import cn from 'clsx';
import { useState } from 'react';
import { twMerge } from 'tailwind-merge';
import { type OptionType } from '../../../../../models/types';

interface SettingsOptionsProps<T> {
  title: string;
  options: OptionType<T>[];
  initialOptionValue: OptionType<T>['value'];
  onSelectOption: (value: OptionType<T>['value']) => void;
  className?: {
    container?: string;
    option?: string;
  };
}

export const SettingsOptions = <T,>({
  title,
  options,
  initialOptionValue,
  onSelectOption,
  className
}: SettingsOptionsProps<T>) => {
  const [selectedOption, setSelectedOption] = useState<T>(initialOptionValue);

  const handleSelectOption = (value: OptionType<T>['value']) => {
    setSelectedOption(value);
    onSelectOption(value);
  };

  return (
    <div className="flex flex-col gap-2">
      <span className="text-center">{title}</span>

      <div className={twMerge('flex flex-col gap-2', className?.container)}>
        {options.map((option) => (
          <button
            key={option.value as string}
            aria-label={option.value as string} //TODO: Improve aria-label for generic options
            onClick={() => handleSelectOption(option.value)}
            className={twMerge(
              'w-full rounded-xl border-2 border-neutral-300 uppercase hover:cursor-pointer dark:border-white/20',
              cn({
                'border-rose-600 bg-rose-100 dark:border-rose-600 dark:bg-rose-950/70':
                  selectedOption === option.value
              }),
              className?.option
            )}
          >
            {option.component}
          </button>
        ))}
      </div>
    </div>
  );
};
