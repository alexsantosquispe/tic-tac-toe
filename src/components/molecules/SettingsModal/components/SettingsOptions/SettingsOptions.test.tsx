import '@testing-library/jest-dom';

import { fireEvent, render, screen } from '@testing-library/react';

import { I18NWrapper } from '../../../../../tests/testsUtils';
import { PLAYER_MODE } from '../../../../../models/types';
import PlayersCard from '../../../../atoms/PlayersCard/PlayersCard';
import { SettingsOptions } from './SettingsOptions';

describe('SettingsOptions', () => {
  const onSelectOptionMock = jest.fn();
  const props = {
    title: 'Player Mode',
    options: [
      { value: PLAYER_MODE.SINGLE_PLAYER, component: <PlayersCard /> },
      {
        value: PLAYER_MODE.TWO_PLAYERS,
        component: <PlayersCard isSinglePlayer={false} />
      }
    ],
    initialOptionValue: PLAYER_MODE.SINGLE_PLAYER,
    onSelectOption: onSelectOptionMock
  };

  it('should render the component correctly', () => {
    const component = render(
      <I18NWrapper>
        <SettingsOptions {...props} />
      </I18NWrapper>
    );

    expect(component).toMatchSnapshot();

    const cardOptions = screen.getAllByRole('button');

    expect(cardOptions).toHaveLength(2);

    expect(cardOptions[0]).toHaveClass(
      'border-rose-600 bg-rose-100 dark:border-rose-600 dark:bg-rose-950/70'
    );
  });

  it('should check the onSelectOption function was executed', () => {
    render(
      <I18NWrapper>
        <SettingsOptions {...props} />
      </I18NWrapper>
    );

    const buttons = screen.getAllByRole('button');

    fireEvent.click(buttons[1]);

    expect(onSelectOptionMock).toHaveBeenCalledTimes(1);

    expect(onSelectOptionMock).toHaveBeenCalledWith(PLAYER_MODE.TWO_PLAYERS);
  });
});
