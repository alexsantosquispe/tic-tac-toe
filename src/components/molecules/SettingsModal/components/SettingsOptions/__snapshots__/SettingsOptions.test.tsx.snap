// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`SettingsOptions should render the component correctly 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="flex flex-col gap-2"
      >
        <span
          class="text-center"
        >
          Player Mode
        </span>
        <div
          class="flex flex-col gap-2"
        >
          <button
            aria-label="singlePlayer"
            class="w-full rounded-xl border-2 uppercase hover:cursor-pointer border-rose-600 bg-rose-100 dark:border-rose-600 dark:bg-rose-950/70"
          >
            <div
              class="flex w-full items-center justify-between gap-8 px-8 py-6"
            >
              <svg
                class="size-12"
                fill="none"
                height="24"
                stroke="currentColor"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
                viewBox="0 0 24 24"
                width="24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M19 21v-2a4 4 0 0 0-4-4H9a4 4 0 0 0-4 4v2"
                />
                <circle
                  cx="12"
                  cy="7"
                  r="4"
                />
              </svg>
              <span
                class="text-xl font-medium"
              >
                VS
              </span>
              <svg
                class="size-12"
                fill="none"
                height="24"
                stroke="currentColor"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
                viewBox="0 0 24 24"
                width="24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M12 8V4H8"
                />
                <rect
                  height="12"
                  rx="2"
                  width="16"
                  x="4"
                  y="8"
                />
                <path
                  d="M2 14h2"
                />
                <path
                  d="M20 14h2"
                />
                <path
                  d="M15 13v2"
                />
                <path
                  d="M9 13v2"
                />
              </svg>
            </div>
          </button>
          <button
            aria-label="twoPlayers"
            class="w-full rounded-xl border-2 border-neutral-300 uppercase hover:cursor-pointer dark:border-white/20"
          >
            <div
              class="flex w-full items-center justify-between gap-8 px-8 py-6"
            >
              <svg
                class="size-12"
                fill="none"
                height="24"
                stroke="currentColor"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
                viewBox="0 0 24 24"
                width="24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M19 21v-2a4 4 0 0 0-4-4H9a4 4 0 0 0-4 4v2"
                />
                <circle
                  cx="12"
                  cy="7"
                  r="4"
                />
              </svg>
              <span
                class="text-xl font-medium"
              >
                VS
              </span>
              <svg
                class="size-12"
                fill="none"
                height="24"
                stroke="currentColor"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
                viewBox="0 0 24 24"
                width="24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M19 21v-2a4 4 0 0 0-4-4H9a4 4 0 0 0-4 4v2"
                />
                <circle
                  cx="12"
                  cy="7"
                  r="4"
                />
              </svg>
            </div>
          </button>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="flex flex-col gap-2"
    >
      <span
        class="text-center"
      >
        Player Mode
      </span>
      <div
        class="flex flex-col gap-2"
      >
        <button
          aria-label="singlePlayer"
          class="w-full rounded-xl border-2 uppercase hover:cursor-pointer border-rose-600 bg-rose-100 dark:border-rose-600 dark:bg-rose-950/70"
        >
          <div
            class="flex w-full items-center justify-between gap-8 px-8 py-6"
          >
            <svg
              class="size-12"
              fill="none"
              height="24"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M19 21v-2a4 4 0 0 0-4-4H9a4 4 0 0 0-4 4v2"
              />
              <circle
                cx="12"
                cy="7"
                r="4"
              />
            </svg>
            <span
              class="text-xl font-medium"
            >
              VS
            </span>
            <svg
              class="size-12"
              fill="none"
              height="24"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M12 8V4H8"
              />
              <rect
                height="12"
                rx="2"
                width="16"
                x="4"
                y="8"
              />
              <path
                d="M2 14h2"
              />
              <path
                d="M20 14h2"
              />
              <path
                d="M15 13v2"
              />
              <path
                d="M9 13v2"
              />
            </svg>
          </div>
        </button>
        <button
          aria-label="twoPlayers"
          class="w-full rounded-xl border-2 border-neutral-300 uppercase hover:cursor-pointer dark:border-white/20"
        >
          <div
            class="flex w-full items-center justify-between gap-8 px-8 py-6"
          >
            <svg
              class="size-12"
              fill="none"
              height="24"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M19 21v-2a4 4 0 0 0-4-4H9a4 4 0 0 0-4 4v2"
              />
              <circle
                cx="12"
                cy="7"
                r="4"
              />
            </svg>
            <span
              class="text-xl font-medium"
            >
              VS
            </span>
            <svg
              class="size-12"
              fill="none"
              height="24"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M19 21v-2a4 4 0 0 0-4-4H9a4 4 0 0 0-4 4v2"
              />
              <circle
                cx="12"
                cy="7"
                r="4"
              />
            </svg>
          </div>
        </button>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
